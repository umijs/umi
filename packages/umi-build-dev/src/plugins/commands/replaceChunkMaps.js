import { join, basename } from 'path';
import { readFileSync, existsSync, writeFileSync } from 'fs';

export const getAssetsManifest = service => {
  const { paths, config } = service;
  const { fileName = 'asset-manifest.json' } = config.manifest;
  const { absOutputPath } = paths;
  // fileName: '../../manifest.json'
  const manifestPath = join(absOutputPath, basename(fileName));
  if (!existsSync(manifestPath)) {
    const errMsg = `Umi SSR error: manifest file ${manifestPath} not found`;
    throw new Error(errMsg);
  }
  const manifest = JSON.parse(readFileSync(manifestPath, 'utf-8'));
  return manifest;
};

export const getServerContent = umiServerPath => {
  if (existsSync(umiServerPath)) {
    const content = readFileSync(umiServerPath, 'utf-8');
    return content;
  }
  return '';
};

export default service => {
  const { paths, config } = service;
  const { absOutputPath } = paths;
  const { basePath = '' } = config.manifest;
  const manifest = getAssetsManifest(service);

  const umiServerPath = join(absOutputPath, 'umi.server.js');
  const umiServer = getServerContent(umiServerPath);
  const result = umiServer
    .replace(/__UMI_SERVER__\.js/g, manifest[`${basePath}umi.js`].split('/').pop())
    .replace(
      /__UMI_SERVER__\.css/g,
      // umi.css may not exist when using dynamic Routing
      manifest[`${basePath}umi.css`] ? manifest[`${basePath}umi.css`].split('/').pop() : '',
    );

  writeFileSync(umiServerPath, result, 'utf-8');
};
