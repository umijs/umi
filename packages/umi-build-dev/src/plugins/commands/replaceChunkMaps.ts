import { join } from 'path';
import { readFileSync, existsSync, writeFileSync } from 'fs';
import { IWebpack, IApi } from 'umi-types';
import { IChunkGroup, getChunkGroupData, isAssetsType } from '../../utils/getChunkMap';

export const getContent = (path: string): string => {
  if (existsSync(path)) {
    const content = readFileSync(path, 'utf-8');
    return content;
  }
  return '';
};

export default (service: IApi, clientStat: IWebpack.Stats) => {
  const { paths } = service;
  const { absOutputPath } = paths;

  const { chunkGroups } = clientStat.compilation;
  const chunkGroupData: IChunkGroup[] = getChunkGroupData(chunkGroups);
  // get umi.js / umi.css
  const { chunks: umiChunk = [] } = chunkGroupData.find(chunk => chunk.name === 'umi') || {};

  const umiJS = umiChunk.find(chunk => isAssetsType('js', chunk)) || '';
  const umiCSS = umiChunk.find(chunk => isAssetsType('css', chunk)) || '';

  const replaceUmiJS = [/__UMI_SERVER__\.js/g, umiJS];
  const replaceUmiCSS = umiCSS
    ? [/__UMI_SERVER__\.css/g, umiCSS]
    : [/,[^,]*createElement\("link".*?__UMI_SERVER__\.css.*?\)/gs, ''];

  const umiServerPath = join(absOutputPath, 'umi.server.js');
  const umiServer = getContent(umiServerPath);

  const result = umiServer.replace(...replaceUmiJS).replace(...replaceUmiCSS);
  writeFileSync(umiServerPath, result, 'utf-8');
};
