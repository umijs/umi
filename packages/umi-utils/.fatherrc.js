export default {
  target: 'node',
  cjs: 'babel',
};
