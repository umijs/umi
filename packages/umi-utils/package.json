{
  "name": "umi-utils",
  "version": "1.3.1",
  "main": "./lib/index.js",
  "repository": {
    "type": "git",
    "url": "http://github.com/umijs/umi/tree/master/packages/umi-utils"
  },
  "homepage": "http://github.com/umijs/umi/tree/master/packages/umi-utils",
  "authors": [
    "chencheng <sorrycc@gmail.com> (https://github.com/sorrycc)"
  ],
  "bugs": {
    "url": "http://github.com/umijs/umi/issues"
  },
  "license": "MIT",
  "files": [
    "lib",
    "src"
  ],
  "dependencies": {
    "is-url": "1.2.4",
    "slash2": "2.0.0"
  }
}
