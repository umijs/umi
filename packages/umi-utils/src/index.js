export { default as findJS } from './findJS';
export { default as compatDirname } from './compatDirname';
export { default as winPath } from './winPath';
