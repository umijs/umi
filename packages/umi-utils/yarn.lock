# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


is-url@1.2.4:
  version "1.2.4"
  resolved "https://registry.yarnpkg.com/is-url/-/is-url-1.2.4.tgz#04a4df46d28c4cff3d73d01ff06abeb318a1aa52"
  integrity sha512-ITvGim8FhRiYe4IQ5uHSkj7pVaPDrCTkNd3yq3cV7iZAcJdHTUMPMEHcqSOy9xZ9qFenQCvi+2wjH9a1nXqHww==

slash2@2.0.0:
  version "2.0.0"
  resolved "https://registry.yarnpkg.com/slash2/-/slash2-2.0.0.tgz#f4e0a11708b8545b912695981cf7096f52c63487"
  integrity sha512-7ElvBydJPi3MHU/KEOblFSbO/skl4Z69jKkFCpYIYVOMSIZsKi4gYU43HGeZPmjxCXrHekoDAAewphPQNnsqtA==
