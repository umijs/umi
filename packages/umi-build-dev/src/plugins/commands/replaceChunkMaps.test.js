import { join, basename } from 'path';
import { winPath } from 'umi-utils';
import replaceChunkMaps from './replaceChunkMaps';

const fs = require('fs');

describe('replaceChunkMaps', () => {
  let result = {};
  let spy;
  let spy2;
  let spy3;

  beforeAll(() => {
    spy = jest.spyOn(fs, 'writeFileSync').mockImplementation((filePath, data) => {
      result[basename(filePath)] = data;
    });
    spy2 = jest.spyOn(fs, 'readFileSync').mockImplementation(
      () => `
      serverRender = async ctx => {
        const htmlTemplateMap = {
          '/': _react.default.createElement("html", null, _react.default.createElement("head", null,_react.default.createElement("link", {
            rel: "stylesheet",
            href: "antd.css"
          }), _react.default.createElement("link", {
            rel: "stylesheet",
            href: "/umiPublic/__UMI_SERVER__.css"
          }), _react.default.createElement("script", {
            src: "/umiPublic/__UMI_SERVER__.js"
          }))),
        };
        return {
          htmlElement: htmlTemplateMap[pathname],
          rootContainer
        };
      }; // using project react-dom version
      // https://github.com/facebook/react/issues/13991


      exports.ReactDOMServer = ReactDOMServer = __webpack_require__(/*! react-dom/server */ "react-dom/server");
      `,
    );
    spy3 = jest.spyOn(fs, 'existsSync').mockImplementation(() => true);
  });
  beforeEach(() => {
    result = {};
  });
  afterEach(() => {
    result = {};
  });
  afterAll(() => {
    spy.mockRestore();
    spy2.mockRestore();
    spy3.mockRestore();
  });

  it('common umi.js, umi.css', () => {
    const service = {
      paths: {
        absOutputPath: join(winPath(__dirname), 'fixtures/chunkMaps'),
      },
      config: {
        ssr: true,
      },
      routes: [
        {
          path: '/',
          routes: [],
        },
      ],
    };
    const clientStat = {
      compilation: {
        chunkGroups: [
          {
            name: 'umi',
            chunks: [
              {
                files: ['umi.baa67d11.css', 'umi.6791e2ab.js'],
              },
            ],
          },
        ],
      },
    };
    replaceChunkMaps(service, clientStat);
    expect(result['umi.server.js']).toMatchSnapshot();
  });

  it('common umi.js without umi.css', () => {
    const service = {
      paths: {
        absOutputPath: join(winPath(__dirname), 'fixtures/chunkMaps'),
      },
      config: {
        ssr: true,
      },
      routes: [
        {
          path: '/',
          routes: [],
        },
      ],
    };
    const clientStat = {
      compilation: {
        chunkGroups: [
          {
            name: 'umi',
            chunks: [
              {
                files: ['umi.6791e2ab.js'],
              },
            ],
          },
        ],
      },
    };
    replaceChunkMaps(service, clientStat);
    expect(result['umi.server.js']).toMatchSnapshot();
  });
});
