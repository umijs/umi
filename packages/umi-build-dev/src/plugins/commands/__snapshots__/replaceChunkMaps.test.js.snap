// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceChunkMaps common umi.js without umi.css 1`] = `
"
      serverRender = async ctx => {
        const htmlTemplateMap = {
          '/': _react.default.createElement(\\"html\\", null, _react.default.createElement(\\"head\\", null, _react.default.createElement(\\"script\\", {
            src: \\"/umiPublic/umi.6791e2ab.js\\"
          }))),
        };
        return {
          htmlElement: htmlTemplateMap[pathname],
          rootContainer
        };
      }; // using project react-dom version
      // https://github.com/facebook/react/issues/13991


      exports.ReactDOMServer = ReactDOMServer = __webpack_require__(/*! react-dom/server */ \\"react-dom/server\\");
      "
`;

exports[`replaceChunkMaps common umi.js, umi.css 1`] = `
"
      serverRender = async ctx => {
        const htmlTemplateMap = {
          '/': _react.default.createElement(\\"html\\", null, _react.default.createElement(\\"head\\", null,_react.default.createElement(\\"link\\", {
            rel: \\"stylesheet\\",
            href: \\"antd.css\\"
          }), _react.default.createElement(\\"link\\", {
            rel: \\"stylesheet\\",
            href: \\"/umiPublic/umi.baa67d11.css\\"
          }), _react.default.createElement(\\"script\\", {
            src: \\"/umiPublic/umi.6791e2ab.js\\"
          }))),
        };
        return {
          htmlElement: htmlTemplateMap[pathname],
          rootContainer
        };
      }; // using project react-dom version
      // https://github.com/facebook/react/issues/13991


      exports.ReactDOMServer = ReactDOMServer = __webpack_require__(/*! react-dom/server */ \\"react-dom/server\\");
      "
`;
